File: submodule/university_8988a097-b948-46f0-a5c5/content/research/hydroponics/analysis/一analysis.py
from pathlib import Path
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re

# --------------------------------------------------
# Configuration
# --------------------------------------------------

EXCEL_FILE = Path.home() / "Downloads" / "Hydroponics.xlsx"

OUTPUT_DIR = Path.home() / "Downloads" / "Output"
OUTPUT_DIR.mkdir(exist_ok=True)

# --------------------------------------------------
# Helper
# --------------------------------------------------

def safe_filename(name):
    return re.sub(r'[<>:"/\\|?*]', "_", str(name))

# --------------------------------------------------
# Read workbook
# --------------------------------------------------
excel = pd.ExcelFile(EXCEL_FILE)

for sheet in excel.sheet_names:
    print(f"Processing {sheet}")

    df = pd.read_excel(EXCEL_FILE, sheet_name=sheet)

    # Convert dates
    if "Date" in df.columns:
        df["Date"] = pd.to_datetime(df["Date"], errors="coerce")

    # Convert numeric columns
    for c in df.columns:
        df[c] = pd.to_numeric(df[c], errors="coerce")

    # --------------------------------------------------
    # Identify treatment columns
    # --------------------------------------------------
    treatment_columns = []
    for col in df.columns:
        name = col.lower()
        
        if any(word in name for word in
               ["fert", "fertilizer",
                "h2o2",
                "scar",
                "treatment",
                "group"]):

            treatment_columns.append(col)

    # --------------------------------------------------
    # Identify response variables
    # --------------------------------------------------
    response_columns = []

    keywords = [
        "stem",
        "leaf",
        "height",
        "root",
        "diameter",
        "mass",
        "weight",
        "length",
        "shoot",
        "growth",
        "biomass"
    ]

    for col in df.columns:
        if pd.api.types.is_numeric_dtype(df[col]):
            if any(k in col.lower() for k in keywords):
                response_columns.append(col)

    if len(response_columns) == 0:
        response_columns = df.select_dtypes(include=np.number).columns.tolist()

    # --------------------------------------------------
    # Summary statistics
    # --------------------------------------------------
    stats = df[response_columns].describe()
    stats.to_csv(OUTPUT_DIR / f"{safe_filename(sheet)}_Statistics.csv")

    # --------------------------------------------------
    # Compare treatments
    # --------------------------------------------------
    for treatment in treatment_columns:

        for response in response_columns:

            subset = df[[treatment, response]].dropna()

            if subset.empty:
                continue

            # -------------------------------
            # Bar chart
            # -------------------------------

            means = subset.groupby(treatment)[response].mean()

            plt.figure(figsize=(7,5))

            means.plot(kind="bar")

            plt.ylabel(response)
            plt.title(f"{response} vs {treatment}")

            plt.tight_layout()

            plt.savefig(OUTPUT_DIR / f"{safe_filename(sheet)}_{safe_filename(response)}_vs_{safe_filename(treatment)}_Bar.png", dpi=200)

            plt.close()

            # -------------------------------
            # Boxplot
            # -------------------------------

            plt.figure(figsize=(7,5))

            subset.boxplot(column=response, by=treatment)

            plt.title(f"{response} by {treatment}")

            plt.suptitle("")

            plt.tight_layout()

            plt.savefig(OUTPUT_DIR / f"{safe_filename(sheet)}_{safe_filename(response)}_vs_{safe_filename(treatment)}_Box.png", dpi=200)

            plt.close()

            # -------------------------------
            # Scatter if treatment numeric
            # -------------------------------

            if pd.api.types.is_numeric_dtype(subset[treatment]):

                plt.figure(figsize=(7,5))

                plt.scatter(subset[treatment], subset[response])

                # Trendline
                z = np.polyfit(subset[treatment], subset[response], 1)

                p = np.poly1d(z)

                x = np.linspace(subset[treatment].min(), subset[treatment].max(), 100)

                plt.plot(x, p(x), linewidth=2)

                plt.xlabel(treatment)
                plt.ylabel(response)

                plt.title(f"{response} vs {treatment}")

                plt.tight_layout()

                plt.savefig(OUTPUT_DIR / f"{safe_filename(sheet)}_{safe_filename(response)}_vs_{safe_filename(treatment)}_Scatter.png", dpi=200)

                plt.close()

    # --------------------------------------------------
    # Time series
    # --------------------------------------------------

    if "Date" in df.columns:

        for treatment in treatment_columns:

            for response in response_columns:

                if treatment not in df.columns:
                    continue

                plot_df = df[["Date", treatment, response]].dropna()

                if plot_df.empty:
                    continue

                plt.figure(figsize=(9,5))

                for group in plot_df[treatment].unique():

                    temp = plot_df[plot_df[treatment] == group]

                    temp = temp.sort_values("Date")

                    plt.plot(temp["Date"], temp[response], marker="o", label=str(group))

                plt.legend(title=treatment)

                plt.xlabel("Date")
                plt.ylabel(response)

                plt.title(f"{response} over Time")

                plt.xticks(rotation=45)

                plt.tight_layout()

                plt.savefig(OUTPUT_DIR / f"{safe_filename(sheet)}_{safe_filename(response)}_{safe_filename(treatment)}_Timeline.png", dpi=200)

                plt.close()

print("Finished!")

File: submodule/university_8988a097-b948-46f0-a5c5/content/research/hydroponics/analysis/一analysis.R
# =====================================================
# Hydroponics Treatment Analysis
# =====================================================

# Install packages if needed
packages <- c("readxl", "ggplot2", "dplyr", "stringr", "lubridate")

new.packages <- packages[!(packages %in% installed.packages()[,"Package"])]

if(length(new.packages))
  install.packages(new.packages)

library(readxl)
library(ggplot2)
library(dplyr)
library(stringr)
library(lubridate)

# =====================================================
# Configuration
# =====================================================

excel_file <- file.path(Sys.getenv("HOME"), "Downloads", "Hydroponics.xlsx")

output_dir <- file.path(Sys.getenv("HOME"), "Downloads", "Output")

if(!dir.exists(output_dir))
  dir.create(output_dir)

safe_filename <- function(x){
  gsub('[<>:"/\\\\|?*]', "_", x)
}

# =====================================================
# Read workbook
# =====================================================

sheets <- excel_sheets(excel_file)

for(sheet in sheets){

  cat("\nProcessing:", sheet, "\n")

  df <- read_excel(excel_file, sheet = sheet)

  # ---------------------------------------------
  # Convert Date
  # ---------------------------------------------

  if("Date" %in% names(df)){
    df$Date <- as.Date(df$Date)
  }

  # ---------------------------------------------
  # Convert numeric columns where possible
  # ---------------------------------------------

  for(col in names(df)){

    suppressWarnings({

      temp <- as.numeric(df[[col]])

      if(sum(!is.na(temp)) > 0)
        df[[col]] <- temp

    })

  }

  # ---------------------------------------------
  # Detect treatment columns
  # ---------------------------------------------

  treatment_cols <- names(df)[
    str_detect(
      tolower(names(df)),
      "fert|fertilizer|h2o2|scar|treatment|group"
    )
  ]

  # ---------------------------------------------
  # Detect response variables
  # ---------------------------------------------

  keywords <- c(
    "stem",
    "leaf",
    "height",
    "diameter",
    "root",
    "mass",
    "weight",
    "length",
    "shoot",
    "growth",
    "biomass"
  )

  response_cols <- c()

  for(col in names(df)){

    if(is.numeric(df[[col]])){

      if(any(str_detect(tolower(col), keywords)))
        response_cols <- c(response_cols, col)

    }

  }

  if(length(response_cols)==0){

    response_cols <- names(df)[sapply(df,is.numeric)]

  }

  # ---------------------------------------------
  # Summary statistics
  # ---------------------------------------------

  stats <- data.frame(
    Variable = response_cols,
    Count = sapply(df[response_cols], function(x) sum(!is.na(x))),
    Mean = sapply(df[response_cols], mean, na.rm=TRUE),
    Median = sapply(df[response_cols], median, na.rm=TRUE),
    SD = sapply(df[response_cols], sd, na.rm=TRUE),
    Minimum = sapply(df[response_cols], min, na.rm=TRUE),
    Maximum = sapply(df[response_cols], max, na.rm=TRUE)
  )

  write.csv(
    stats,
    file.path(output_dir,
              paste0(safe_filename(sheet), "_Statistics.csv")),
    row.names = FALSE
  )

  # ---------------------------------------------
  # Treatment comparisons
  # ---------------------------------------------

  for(treatment in treatment_cols){

    for(response in response_cols){

      plot_df <- df %>%
        select(all_of(c(treatment,response))) %>%
        na.omit()

      if(nrow(plot_df)==0)
        next

      # ---------------------------------
      # Bar chart
      # ---------------------------------

      means <- plot_df %>%
        group_by(.data[[treatment]]) %>%
        summarise(Mean = mean(.data[[response]]),
                  .groups="drop")

      p <- ggplot(means,
                  aes(x=.data[[treatment]],
                      y=Mean))+
        geom_col()+
        labs(
          title=paste(response,"vs",treatment),
          x=treatment,
          y=response
        )+
        theme_bw()

      ggsave(
        file.path(
          output_dir,
          paste0(
            safe_filename(sheet),
            "_",
            safe_filename(response),
            "_vs_",
            safe_filename(treatment),
            "_Bar.png"
          )
        ),
        p,
        width=7,
        height=5
      )

      # ---------------------------------
      # Boxplot
      # ---------------------------------

      p <- ggplot(plot_df,
                  aes(x=factor(.data[[treatment]]),
                      y=.data[[response]]))+
        geom_boxplot()+
        labs(
          title=paste(response,"by",treatment),
          x=treatment,
          y=response
        )+
        theme_bw()

      ggsave(
        file.path(
          output_dir,
          paste0(
            safe_filename(sheet),
            "_",
            safe_filename(response),
            "_vs_",
            safe_filename(treatment),
            "_Boxplot.png"
          )
        ),
        p,
        width=7,
        height=5
      )

      # ---------------------------------
      # Scatter plot
      # ---------------------------------

      if(is.numeric(plot_df[[treatment]])){

        p <- ggplot(plot_df,
                    aes(x=.data[[treatment]],
                        y=.data[[response]]))+
          geom_point(size=3)+
          geom_smooth(method="lm",
                      se=FALSE)+
          labs(
            title=paste(response,"vs",treatment),
            x=treatment,
            y=response
          )+
          theme_bw()

        ggsave(
          file.path(
            output_dir,
            paste0(
              safe_filename(sheet),
              "_",
              safe_filename(response),
              "_vs_",
              safe_filename(treatment),
              "_Scatter.png"
            )
          ),
          p,
          width=7,
          height=5
        )

      }

    }

  }

  # ---------------------------------------------
  # Time series
  # ---------------------------------------------

  if("Date" %in% names(df)){

    for(treatment in treatment_cols){

      for(response in response_cols){

        plot_df <- df %>%
          select(Date,
                 all_of(treatment),
                 all_of(response)) %>%
          na.omit()

        if(nrow(plot_df)==0)
          next

        p <- ggplot(
          plot_df,
          aes(
            x=Date,
            y=.data[[response]],
            colour=factor(.data[[treatment]])
          )
        )+
          geom_line()+
          geom_point()+
          labs(
            title=paste(response,"over Time"),
            colour=treatment,
            y=response
          )+
          theme_bw()

        ggsave(
          file.path(
            output_dir,
            paste0(
              safe_filename(sheet),
              "_",
              safe_filename(response),
              "_",
              safe_filename(treatment),
              "_Timeline.png"
            )
          ),
          p,
          width=8,
          height=5
        )

      }

    }

  }

}

cat("\nAnalysis Complete!\n")